File: cpp/valid_anagram.cpp
class Solution {
public:
    // optimal letter frequency approach
    bool isAnagram(string s, string t) {
        if (s.size() != t.size()) {
            return false;
        }

        int letterFreqs[26] = { 0 };
        for (char &letter : s) {
            ++letterFreqs[int(letter) - int('a')];
        }

        for (char &letter : t) {
            --letterFreqs[int(letter) - int('a')];
            // word t had more of current letter than word s
            if (letterFreqs[int(letter) - int('a')] < 0) {
                return false;
            }
        }

        // length of t and s is same up to here, do not need to check if
        // t had less letters than s
        return true;    
    }
};

File: python/valid_anagram.py
class Solution:
    def isAnagram(self, s: str, t: str) -> bool:
        charsByCount = {}

        if len(s) != len(t):
            return False

        # dump s's characters in map with count
        for char in s:
            if char not in charsByCount:
                charsByCount[char] = 1
            else:
                charsByCount[char] += 1

        # decrement count by all characters in t
        for char in t:
            if char not in charsByCount:
                return False

            charsByCount[char] -= 1
            if charsByCount[char] < 0:
                return False

        # length of t and s is same up to here, do not need to check if
        # t had less letters than s (any letter frequencies >= 1)
        return True
